homeassistant/components/utility_meter/__init__.py

```diff
@@ -230,22 +230,18 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
 
 async def config_entry_update_listener(hass: HomeAssistant, entry: ConfigEntry) -> None:
     """Update listener, called when the config entry options are changed."""
-    old_source = hass.data[DATA_UTILITY][entry.entry_id]["source"]
-    await hass.config_entries.async_reload(entry.entry_id)
-
-    if old_source == entry.options[CONF_SOURCE_SENSOR]:
-        return
 
-    entity_registry = er.async_get(hass)
+    # Remove device link for entry, the source device may have changed.
+    # The link will be recreated after load.
     device_registry = dr.async_get(hass)
+    devices = device_registry.devices.get_devices_for_config_entry_id(entry.entry_id)
 
-    old_source_entity = entity_registry.async_get(old_source)
-    if not old_source_entity or not old_source_entity.device_id:
-        return
+    for device in devices:
+        device_registry.async_update_device(
+            device.id, remove_config_entry_id=entry.entry_id
+        )
```

Comment:
I'm not convinced the unconditional remove here is great, I think we should add a check in the for loop if the device has an entity matching our source sensor before continuing.

Reply:
The only point of the integration is links to another entity. Anything linking to to this config entry that is not part of its source config is wrong.

We could avoid removing the link if it is still valid, but it adds complexity which i couldn't find a good reason for.

If we do validate the links, i think the consideration of doing it on startup instead make a lot of sense.

Reply:
> Anything linking to to this config entry that is not part of its source config is wrong.

Yeah, exactly, but the PR removes ALL links, not just the links which are not part of the source config.